### ISSUE TITLE
[Coordination] Descriptive failure clustering; event-table implementation in #7386

### ISSUE THREAD

**Issue by ll7** [ll7/robot_sf_ll7#7396]
## Current issue-planning review — 2026-09-15

**Audit disposition:** `blocked`. The current audit classifies the issue as blocked; its prerequisite, evidence, or triage gate is incomplete.

**First action:** Build a reproducible event-level data and clustering pipeline that groups similar Robot SF scenarios and failures, exposes outliers and rare cases, and supports expert naming and retrieval. Add a separate, explicitly controlled root-cause experiment track.

### Current bounded plan

1. Inspect the named owners and dependencies before editing; keep the live issue scope and fail-closed evidence boundary intact.
2. Run the named focused validation, record exact output and remaining blockers, and keep fallback, degraded, unavailable, and inconclusive outcomes explicit.

### Files and owners to inspect

- No repository path was identified in the live body; establish the canonical owner before editing.

### Acceptance and validation

- Demonstrate the stated outcome from the live issue with a focused, reproducible check.
- Keep any missing prerequisite or inconclusive result explicit rather than treating it as success.

### Current dependencies and issue-text corrections

- No supplied individual draft existed for this issue in the handoff package; this is a conservative plan derived from the refreshed live body, and the historical issue text is preserved below.
- Live issue references to verify: https://github.com/ll7/robot_sf_ll7/issues/7381; https://github.com/ll7/robot_sf_ll7/issues/7386; https://github.com/ll7/robot_sf_ll7/issues/191; https://github.com/ll7/robot_sf_ll7/issues/2754
- Keep current out-of-scope, ownership, provenance, rights, compute, and claim boundaries intact. Do not close the issue merely because this plan was posted.

### Lifecycle and claim boundary

- Current state labels: `state:blocked`, `state:parked`.
- The issue remains open. No closure, compute submission, release/publication, evidence deletion, or scientific-claim promotion is authorized by this review.
- This review does not request a lifecycle transition; existing assignees, ownership, valid priority labels, and unrelated labels are preserved.
- Proposed non-lifecycle label additions applied with this review: `none`.

**Evidence note:** Full live issue body and complete available comment thread were reviewed during the 2026-09-15 refresh. This is planning metadata, not implementation or benchmark evidence; historical measurements remain historical.

Parent programme: #7381

Deep research: `ll7/diss#1110`

Blocking design decision: #7386

Existing taxonomy baseline: #191

Existing failure-analysis surface: #2754

## Goal

Build a reproducible event-level data and clustering pipeline that groups similar Robot SF scenarios and failures, exposes outliers and rare cases, and supports expert naming and retrieval. Add a separate, explicitly controlled root-cause experiment track.

## Non-goals

- treating clusters as natural ground-truth categories;
- converting a cluster label into a causal diagnosis;
- discarding HDBSCAN noise or rare events;
- replacing the existing operational taxonomy;
- choosing a method from a visually attractive two-dimensional projection;
- using test data to select features, cluster count, or hyperparameters.

### 1. Stable event unit and feature table

Build the event-centred schema selected in #7386. Every row links to parent run, scenario, seed, planner, actor set, event interval, source artifact digests, and manual taxonomy fields.

Feature groups should include geometry, timing, criticality, collision/contact, robot control, planner state, pedestrian dynamics, uncertainty, and event duration/recovery. Record unit, transform, missing-data status, and whether the feature is planner-visible or post-hoc.

### 2. Dataset and split manifest

Use grouped partitions that keep scenario templates, seed lineages, snapshot parents, and near-duplicates together. Fit scaling, imputation, and representation transforms only on training data. Freeze a versioned evaluation manifest.

### 3. Baseline clustering suite

Implement through scikit-learn where possible:

- existing rule-based/manual taxonomy comparison;
- k-means;
- agglomerative clustering;
- HDBSCAN with explicit noise and membership strength.

Use deterministic seeds and store complete preprocessing/model configurations.

### 4. Stability and interpretation

Bootstrap/resample and repeat clustering. Report adjusted Rand index or equivalent agreement, variation across seeds/features, cluster size, noise coverage, representative and boundary cases, and feature distributions.

Create an expert-review packet with blinded samples, proposed descriptive labels, confidence, disagreements, and retained unknowns.

### 5. Optional representation extensions

After baseline completion, evaluate fixed-length time-series features, trajectory distances, TS2Vec-style embeddings, or interaction graphs. These must use the same split and stability protocol and must demonstrate additional downstream utility.

### 6. Root-cause track

Create known injected mechanisms and controlled counterfactual/replay tests. Store candidate cause, assumed causal graph or intervention, supporting and contradicting evidence, and result. Do not populate a `cause` field from clustering alone.

## Work packages

- [ ] Resolve #7386 and freeze the unit of analysis and feature schema.
- [ ] Map #191 taxonomy fields into the event table without changing their meaning.
- [ ] Integrate existing failure/event extraction from #2754.
- [ ] Implement versioned feature registry with units and provenance.
- [ ] Implement grouped split and near-duplicate audit.
- [ ] Implement missing-data and scaling pipeline fitted on training data only.
- [ ] Implement k-means, agglomerative, and HDBSCAN baselines.
- [ ] Implement stability, noise, and rare-case reporting.
- [ ] Implement representative/boundary/outlier retrieval.
- [ ] Implement expert-review and semantic-label records.
- [ ] Add optional time-series/trajectory representation benchmark.
- [ ] Build synthetic known-mechanism fixtures.
- [ ] Implement separate counterfactual/root-cause result schema.
- [ ] Add cluster browser/export for the visual workbench.
- [ ] Add publication tables/figures with uncertainty and limitations.

## Required artifacts

- event schema and feature registry;
- source and split manifests;
- duplicate/leakage audit;
- fitted preprocessing objects and configurations;
- cluster assignments, noise status, membership/confidence, and model digest;
- stability and sensitivity report;
- representative/boundary/outlier case lists;
- expert-review and naming record;
- root-cause experiment records kept separate from clustering;
- source/environment provenance.

### Unit tests

- feature units and transforms;
- missing-data semantics;
- grouped split invariants;
- preprocessing fit isolation;
- deterministic baseline configuration;
- noise and unknown preservation;
- manual taxonomy round trip.

### Integration tests

- known synthetic clusters are recovered within declared tolerance;
- injected rare cases remain retrievable when marked noise;
- duplicate lineage cannot cross splits;
- changing a visualization projection does not change stored cluster assignments;
- causal fields remain empty without a controlled root-cause record;
- every assignment links back to event evidence.

### Regression tests

- #191 taxonomy outputs remain available;
- existing failure-pack/report paths continue to work;
- adding new feature versions does not reinterpret old model artifacts.

## Evaluation matrix

Compare rule-based taxonomy, k-means, agglomerative, and HDBSCAN across feature subsets and bootstrap samples. Report stability, noise coverage, cluster size, expert coherence, known-mechanism recovery, retrieval utility, and sensitivity to scaling/missing data.

## Acceptance criteria

- [ ] #7386 contains a complete author decision.
- [ ] Event rows and model outputs have full source provenance.
- [ ] Manual taxonomy remains a baseline and is not silently overwritten.
- [ ] HDBSCAN noise and rare cases remain first-class retrievable outputs.
- [ ] Stability and expert-review evidence accompany semantic names.
- [ ] No cluster is represented as a root cause without a separate controlled record.
- [ ] Optional learned representations outperform or complement interpretable baselines under the same protocol.

## Claim boundary

The pipeline discovers representation-dependent descriptive groupings. It does not prove discrete natural classes or causes. Root-cause language is limited to separately identified controlled evidence.

```yaml
schema: goal_autopilot_preparation.v1
repository: ll7/robot_sf_ll7
issue: 7396
source_body_sha256: f26e7cf0aa634e01d618352a6cd523f49beeeff08d781b3651524f23632e0ef2
source_comments_sha256: 
audit_schema: open_issue_contract_audit.v1
audit_digest: 1c191411793929640684e0827d763f3a2009a8682ab98dcefa8c2ded3e54f4ea
audit_classification: state_conflict
next_action: reconcile_state_labels
authority: lifecycle_owner
execution_mode: lifecycle
preferred_worker: MaxRunner
expected_pr_runner_label: runner:max
implementation_admitted: False
state_ready_change_proposed: False
mutation_batch: open-issues-20260830-04
```

This packet is preparation evidence only. It never overrides live labels, exact claim state, branch state, typed dependencies, domain gates, compute authority, release authority, or scientific evidence rules.

**Comment by ll7** [comment]
## Portfolio disposition — 2026-08-18

Parked as post-dissertation research. Existing owners remain #191 for the operational taxonomy and #2754 for failure/event extraction and review.

On revival, this issue may add only a **descriptive event-clustering layer**. It must preserve the manual taxonomy as a baseline, retain rare/noise cases, use group-aware splits, and keep semantic naming under expert review.

A cluster label must never populate a causal/root-cause field. Controlled replay or an explicit identified causal model is required for any stronger mechanism statement. The first packet must inventory reusable #191/#2754 fields and prove that no parallel event taxonomy or failure store is introduced.

No model fitting, campaign, or dissertation claim is authorized before parent #7381 explicitly selects the scenario-intelligence product.

**Comment by ll7** [comment]
## Background research — event-table and clustering evaluation implementation

**Recommendation:** make the first implementation an auditable feature-table adapter with leakage tests and a small comparison report. Preserve #191's manual taxonomy and reuse #2754's failure extraction. The method-level findings are posted in #7386; this comment specifies how to make the evaluation executable.

### Source-backed implementation constraints

GroupKFold separates samples by supplied groups [1], but the application must construct groups that actually contain related observations. Scikit-learn's leakage guidance requires fitting preprocessing on training data only [2]. Silhouette is defined only when the number of labels is between two and `n_samples - 1` [3]. The HDBSCAN API has no ordinary new-point `predict` method [4]. These details materially affect the proposed grouped, held-out stability study.

### Proposed bounded implementation

**1. Build the lineage-aware table.** Each row needs an event ID, parent run/scenario/seed/planner/actor set, source ranges, feature-definition version, units, availability mask and source digest. Construct split groups from shared parent lineage, including overlapping windows, snapshot siblings and near-identical perturbations where those would leak the same event. Holding out families tests a different question from holding out seeds within known families; name the chosen question.

Keep identifiers and the manual taxonomy outside the clustering feature matrix. Keep post-hoc outcome descriptors separate from features advertised as available before failure. Detect duplicated row IDs and conflicting provenance as errors.

**2. Freeze preprocessing and controls.** Use a versioned feature list, development-only imputation/scaling and explicit random seeds/initializations. Start with manual labels, k-means and agglomerative clustering as ruled. Do not optimize every metric over an unbounded configuration sweep. Include a small feature ablation to expose dependence on planner identity proxies, missingness or a dominant metric family.

**3. Declare out-of-sample semantics.** K-means can assign to frozen centers. For agglomerative clustering or HDBSCAN, either perform a clearly labelled descriptive clustering of the declared corpus, or add a separately specified frozen-prototype/medoid assignment with an unknown-distance rule. That adapter is not native HDBSCAN prediction. Refitting clusters jointly on the test set answers a transductive descriptive question, not untouched inductive generalization.

**4. Implement meaningful resampling.** Keep the ruled 100 deterministic resamples, but resample parent groups and retain multiplicities and exact seeds. Duplicate bootstrap observations can alter density-based clustering, so document this effect and consider a labelled grouped-subsampling sensitivity check. Compare partitions on matched unique event IDs; report coverage and the treatment of noise alongside agreement. Do not equate missing-data rows with genuine rare-event noise.

**5. Produce a reviewable report.** For each configuration, show cluster sizes, usable-data coverage, noise/unknown fraction, stability distribution, manual-taxonomy association, and medoid/boundary/rare trace examples. Report silhouette as unavailable for all-noise or one-cluster results after the declared filtering. A visually clean 2D projection is an illustration, not the distance space or validation criterion unless explicitly designed that way.

### Acceptance fixtures and remaining evidence

Test that related windows cannot cross splits; test-set changes do not refit the scaler; row permutations do not break joins; all-noise/missing/one-cluster cases remain explicit; and descriptions cannot populate root-cause fields. Then evaluate on the available admitted corpus. No cluster fitting or performance measurement was performed for this comment.

### Primary sources

[1] GroupKFold: https://scikit-learn.org/stable/modules/generated/sklearn.model_selection.GroupKFold.html

[2] Leakage and preprocessing pitfalls: https://scikit-learn.org/stable/common_pitfalls.html

[3] Silhouette definition and valid domain: https://scikit-learn.org/stable/modules/generated/sklearn.metrics.silhouette_score.html

[4] HDBSCAN API: https://scikit-learn.org/stable/modules/generated/sklearn.cluster.HDBSCAN.html

Research checked 2026-09-16. No new human naming gate or change to existing ownership is proposed.

**Comment by ll7** [comment]
## Research-to-implementation handoff — 2026-09-16

The methodological background is complete. Removed `needs-research`, stale `needs-triage`, and `state:parked` from this coordination parent.

**Implement #7386 first.** It is now a bounded `state:ready` implementation leaf for the event-table adapter, feature registry, source provenance, leakage-safe grouped splits, and training-only preprocessing tests. Its updated body contains the selected scope, concrete owners, outputs, and acceptance commands; it requires no further literature review.

This parent retains the broader clustering comparison, stability evaluation, descriptive naming, and optional-method roadmap. It waits for implementation/evaluation deliverables, not a new method-choice decision. #7386 does not depend on this parent closing, and this parent must not be dispatched as a competing implementation of the same adapter.

Existing research boundaries remain: clusters are descriptions, not causes; rare/noise cases remain visible; held-out data cannot tune preprocessing or methods. Fixture-tested software is not corpus-level stability evidence. The current readiness decision is manual triage under the maintainer's request, with normal worker admission/claim and PR checks still required.
